proposals/TP3-xds-error-propagation.md

```diff
@@ -0,0 +1,122 @@
+
+### Related Proposals:
+
+N/A
+
+## Proposal
+
+The proposal introduces two new fields to the DiscoveryResponse proto. The first, `resource_errors`, will detail Resource-specific errors and be included in both SotW and Incremental xDS protocols. The second, `removed_resources`, will specifically identify resources that no longer exist, addressing a current gap in the SotW protocol's signaling capabilities. 
+
+```textproto
+// New Proto Message
+// Contains detailed error detail for a single resource
+message ResourceError {
+ ResourceName resource_name = 1;
+
+ google.rpc.Status error_detail = 2;
+}
+
+message DiscoveryResponse {
+
+// The version of the response data.
+string version_info = 1;
+
+….
+…
+
+
+ // The control plane instance that sent the response.
+  config.core.v3.ControlPlane control_plane = 6;
+
+// NEW_FIELD
+// An optional repeated list of error details that the control plane 
+// can provide to notify the client of issues for the resources that 
+// are subscribed.
+// This allows the xDS management server to provide optional 
+// notifications in case of unavailability, permissions errors 
+// without the client having to wait for the `config fetch` timeout.
+repeated ResourceError resource_errors = N;
+
+// NEW FIELD
+// For non-existing resources.
+repeated ResourceName removed_resources = N;
+}
+
+message DeltaDiscoveryResponse {
+  …
+
+  // NEW FIELD
+  repeated ResourceError resource_errors = N;
+}
+```
+
+### Protocol Behavior
+The client must use this additional field to obtain notification for resources the xDS Management server couldn’t procure. The xDS client should cancel any resource timers once this message is received and convey the error message to the application. With the addition of this field, when a client receives an explicit error or does-not-exist indicator from the management server, it should react the same way it would have if its does-not-exist timer fired. 
+
+The xDS Management server is only expected to return the error message once rather than throughout for future responses. The client is expected to remember the error message until either a new error message is returned or the resource is returned. 
+
+### Wildcard Resources
+
+It is possible to subscribe to all resources by a client using a wildcard or "" resource name. The control plane in this case can provide error details for two different use cases. One when the issue is with the glob itself or later on when the issue is specific to individual resources. 
+
+1. For errors associated with wildcard("" for legacy or "*") and for xDS-TP glob collections; the control plane `error_details` resource name will match the relevant wildcard request("" or "*" or xDS-TP glob collections). This can be used by the control plane to indicate an error with the collection as a whole.
```

Comment:
My inclination is to say that we support this case only for glob collections, not for old-style wildcard subscriptions.  We want people to migrate away from old-style wildcard subscriptions anyway.

Reply:
I don't have a strong opinion about supporting old-style wildcard, @adisuissa any preferences?

Reply:
@adisuissa, any concern about not supporting old-style wildcard subscriptions here?

Reply:
While Envoy is configured to use "*" for wildcard, it doesn't, as this may break current servers. It may have been possible if there was a plan to upgrade the protocol to v4.
Specifically for this field, it would be possible to have it asymmetric, that is Envoy subscribes using the old wildcard (""), and receives a response where the resource_name in the resource_errors is the new wildcard ("*").

Reply:
The intent for introducing `*` was that Envoy would use `*` for wildcard if it was *also* subscribing to on-demand resources.  Envoy was not going to do that if it was *not* subscribing to on-demand resources, since that preserved backward-compatibility for the common case with existing control plane implementations.  See discussion in https://github.com/envoyproxy/envoy/pull/16861.

I agree that we could support wildcard subscriptions here, and always using `*` here would be one way to do that.  My question here is really about whether that's really necessary in the first place.  It seems simpler to just say that this mechanism doesn't support wildcard subscriptions, and if you want to use this, you need to upgrade to xdstp collections (either list or glob).  Is there any reason that won't work?  Do we actually have any cases where we'd need this with wildcard subscriptions?

Reply:
@adisuissa, still need your input on this question about wildcard subscriptions.  (You might have missed it on your last review pass, since this thread is on an earlier commit, and the github UI is terrible...)

Reply:
As discussed offline, it may be better to have this intentionally saying that the mechanism doesn't support wildcard subscriptions at the moment (and in the future this may be modified).

Reply:
Done, change the contents here to only refer to glob collections.